libxmf/XmfVpxEncoder.c

```diff
@@ -0,0 +1,198 @@
+#include "XmfVpxEncoder.h"
+#include <stdlib.h>
+#include <string.h>
+#include <vpx/vpx_encoder.h>
+#include <vpx/vp8cx.h>
+#include "XmfVpxPacket.h"
+
+struct xmf_vpx_encoder
+{
+	vpx_codec_ctx_t codec;
+	vpx_codec_enc_cfg_t cfg;
+	vpx_codec_pts_t pts;
+	XmfVpxEncoderError lastError;
+};
+
+XmfVpxEncoder *XmfVpxEncoder_Create(XmfVpxEncoderConfig config)
+{
+	XmfVpxEncoder *encoder = (XmfVpxEncoder *)malloc(sizeof(XmfVpxEncoder));
+	if (!encoder)
+	{
+		return NULL;
+	}
+
+	vpx_codec_iface_t *iface = NULL;
+	if (config.codec == VP8)
+	{
+		iface = vpx_codec_vp8_cx();
+	}
+	else if (config.codec == VP9)
+	{
+		iface = vpx_codec_vp9_cx();
+	}
+	else
+	{
+		free(encoder);
+		return NULL; // Invalid codec parameter
+	}
+
+	vpx_codec_err_t res = vpx_codec_enc_config_default(iface, &encoder->cfg, 0);
+	if (res != VPX_CODEC_OK)
+	{
+		free(encoder);
+		return NULL; // Failed to get default config
+	}
+
+	// Set encoding parameters
+	encoder->cfg.g_w = config.width;
+	encoder->cfg.g_h = config.height;
+	encoder->cfg.rc_target_bitrate = config.bitrate;
+	encoder->cfg.g_timebase.num = config.timebase_num;
+	encoder->cfg.g_timebase.den = config.timebase_den;
+	encoder->cfg.g_threads = config.threads;
+	encoder->cfg.g_error_resilient = 1;
+
+	// allow keyframes at any time
+	encoder->cfg.kf_mode = VPX_KF_AUTO;
+	encoder->cfg.kf_min_dist = 0;
+	encoder->cfg.kf_max_dist = 9999;
+
+	// Initialize codec
+	res = vpx_codec_enc_init(&encoder->codec, iface, &encoder->cfg, 0);
+	if (res != VPX_CODEC_OK)
+	{
+		free(encoder);
+		return NULL; // Failed to initialize codec
+	}
+
+	encoder->pts = 0;
+	encoder->lastError.code = NO_ERROR;
+
+	return encoder;
+}
+
+int XmfVpxEncoder_EncodeFrame(XmfVpxEncoder *encoder, const XmfVpxImage *image, vpx_codec_pts_t pts, unsigned long duration, unsigned int flags)
+{
+	if (!encoder || !image)
+	{
+		return -1;
+	}
+
+	vpx_image_t *vpx_image = XmfVpxImage_GetData(image);
+
+	if (!vpx_image)
+	{
+		encoder->lastError.code = INVALID_PARAM;
+		return -1;
+	}
+
+	vpx_codec_err_t res = vpx_codec_encode(&encoder->codec, vpx_image, pts, duration, flags, VPX_DL_REALTIME);
+	if (res != VPX_CODEC_OK)
+	{
+		encoder->lastError.code = VPX_ERROR;
+		encoder->lastError.detail.vpx_error.error_code = res;
+		return -1;
+	}
+	// Reset the iterator after encoding
```

Comment:
What iterator? Seems like an invalid comment (doesn't make sense in context)

Reply:
updated